compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -1320,17 +1320,40 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 ty::PredicateKind::Clause(clause) => match clause {
                     ty::Clause::Trait(trait_pred) => {
                         assert_eq!(trait_pred.polarity, ty::ImplPolarity::Positive);
-                        trait_bounds.push((
-                            bound_pred.rebind(trait_pred.trait_ref),
-                            span,
-                            trait_pred.constness,
-                        ));
+                        if let ty::Alias(ty::Projection, projection_ty) =
+                            *trait_pred.trait_ref.self_ty().kind()
+                        {
+                            // This is a bad `dyn T<Item: Bound>`
+                            projection_bounds.push((
+                                bound_pred.rebind(ty::ProjectionPredicate {
+                                    projection_ty,
+                                    term: tcx.ty_error().into(),
+                                }),
+                                span,
+                            ));
```

Comment:
Can lowering set the constraint to be `Item = TyKind::Err`, and avoid the special case here?